### PR TITLE
support multi-gpu; fix inference duration queue bug

### PR DESCRIPTION
**Pull request by ylwu-amzn**
Signed-off-by: Yaliang Wu <ylwu@amazon.com>

### Description
support multi-gpu; fix inference duration queue bug

### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dhrubo-os — APPROVED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java [hunk 1]
```diff
@@ -170,39 +177,64 @@ protected void loadTextEmbeddingModel(File modelZipFile, String modelId, String
...
-                    if (ONNX_ENGINE.equals(engine)) { //ONNX
-                        criteriaBuilder.optTranslator(new ONNXSentenceTransformerTextEmbeddingTranslator());
-                    } else { // pytorch
-                        if (transformersType == SENTENCE_TRANSFORMERS) {
-                            criteriaBuilder.optTranslator(new SentenceTransformerTextEmbeddingTranslator());
-                        } else {
-                            criteriaBuilder.optTranslatorFactory(new HuggingfaceTextEmbeddingTranslatorFactory());
-                        }
+                    devices = Engine.getInstance().getDevices();
+                    if (devices == null || devices.length == 0) {
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java:181]: Cool, will remove this checking.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java [hunk 2]
```diff
@@ -170,39 +177,64 @@ protected void loadTextEmbeddingModel(File modelZipFile, String modelId, String
...
-                    TextEmbeddingModelConfig.FrameworkType transformersType = textEmbeddingModelConfig.getFrameworkType();
-                    if (ONNX_ENGINE.equals(engine)) { //ONNX
-                        criteriaBuilder.optTranslator(new ONNXSentenceTransformerTextEmbeddingTranslator());
-                    } else { // pytorch
-                        if (transformersType == SENTENCE_TRANSFORMERS) {
-                            criteriaBuilder.optTranslator(new SentenceTransformerTextEmbeddingTranslator());
-                        } else {
-                            criteriaBuilder.optTranslatorFactory(new HuggingfaceTextEmbeddingTranslatorFactory());
-                        }
+                    devices = Engine.getInstance().getDevices();
```
[the remark this answers is not in the document]
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java:180]: Sure, will change this part. BTW, here we run `Engine.getInstance()` just to get devices. Will it return different devices for different engines?

[the remark this answers is not in the document]
Comment (frankfliu) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java:180]: Right, not all engine support GPU.

[the remark this answers is not in the document]
Comment (Zhangxunmt) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java:180]: I may lack some context here. Does this Engine/device thing relate to GPU, etc? Is there any public documentations for this ai.djl?

[the remark this answers is not in the document]
Comment (frankfliu) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java:180]: `CudaUtils.getGpuCount()` always return number of physical GPUS.
The `Engine.getDevices()` API relies on the following:
1. If the engine support GPU or not (for example: LightGBM doesn't support GPU)
2. If the engine native library build with GPU enabled (pytorch-native-cpu build doesn't support GPU)
3. If the CUDA version is supported by the engine (CUDA 9.2 is no longer supported)
4. If the CUDA [compute capability](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capabilities) is supported by the engine (e.g.  is not built into MXNet 1.8.0)

[the remark this answers is not in the document]
Comment (jngz-es) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java]: LightGBM supports GPU with OpenCL, and experimentally supports CUDA, right?

[the remark this answers is not in the document]
Comment (frankfliu) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java]: DJL doesn't have LightGBM CUDA support. call `Engine.hasCapability(StandardCapabilities.CUDA)` to check if the engine has CUDA support.

**Review by Zhangxunmt — DISMISSED** [review]

File: plugin/src/main/java/org/opensearch/ml/model/MLModelCache.java [hunk 3]
```diff
@@ -76,17 +76,34 @@ public void clear() {
...
+        resizeInferenceQueue(maxRequestCount, predictRequestDurationQueue);
+        resizeInferenceQueue(maxRequestCount, modelInferenceDurationQueue);
+    }
+
+    private void resizeInferenceQueue(long maxRequestCount, Queue<Double> queue) {
+        if (maxRequestCount <= 0) {
+            queue.clear();
+        } else {
+            while (queue.size() >= maxRequestCount) {
+                queue.poll();
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/model/MLModelCache.java:104]: Just curious to know what happens when we poll the queue?

File: plugin/src/main/java/org/opensearch/ml/model/MLModelCache.java [hunk 4]
```diff
@@ -76,17 +76,34 @@ public void clear() {
...
+        }
+    }
+
+    public void resizeMonitoringQueue(long maxRequestCount) {
+        log.debug("resize inference duration monitoring queue with size {}", maxRequestCount);
+        resizeInferenceQueue(maxRequestCount, predictRequestDurationQueue);
+        resizeInferenceQueue(maxRequestCount, modelInferenceDurationQueue);
+    }
+
+    private void resizeInferenceQueue(long maxRequestCount, Queue<Double> queue) {
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/model/MLModelCache.java:99]: may be the change the name as this is also resizing `predictRequestDurationQueue`?

**Review by dhrubo-os — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/model/MLModelCache.java (hunk 3 above)
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/model/MLModelCache.java:104]: This is mainly for testing and monitoring, it won't impact the main function, just make sure we don't store too much data in queue. When poll the queue, it just remove old items.

File: plugin/src/main/java/org/opensearch/ml/model/MLModelCache.java (hunk 4 above)
Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/model/MLModelCache.java:99]: Both `predictRequestDurationQueue` and `modelInferenceDurationQueue` are for inferencing, I think should be ok.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java [hunk 5]
```diff
@@ -170,39 +177,61 @@ protected void loadTextEmbeddingModel(File modelZipFile, String modelId, String
...
                     return null;
                 } catch (Exception e) {
                     String errorMessage = "Failed to load model " + modelId;
                     log.error(errorMessage, e);
                     close();
+                    if (predictorList.size() > 0) {
+                        closePredictors(predictorList.toArray(new Predictor[0]));
+                        predictorList.clear();
+                    }
+                    if (modelList.size() > 0) {
```
Comment (jngz-es) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java:231]: If model list > 0, should we continue with the models loaded successfully?

**Review by Zhangxunmt — APPROVED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java [hunk 6]
```diff
@@ -216,12 +245,30 @@ protected void loadTextEmbeddingModel(File modelZipFile, String modelId, String
...
-        log.debug("start to predict text embedding model {}", modelId);
         try {
             return AccessController.doPrivileged((PrivilegedExceptionAction<ModelTensorOutput>) () -> {
                 Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
-                if (predictor == null) {
+                int currentDevice = nextDevice.getAndIncrement();
+                if (currentDevice > devices.length - 1) {
+                    currentDevice = currentDevice % devices.length;
+                    nextDevice.set(currentDevice + 1);
+                }
```
Comment (jngz-es) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java:270]: I think this whole section are critical needed to thread safe. For example, device length is 2, we could have two threads. Thread A has current 2, thread B has current 3. Then thread B run before thread A, then nextDevice will be set wrong number. If we are concerned the performance, how about just 
int currentDevice = nextDevice.getAndIncrement() % devices.length;

File: plugin/src/main/java/org/opensearch/ml/cluster/MLCommonsClusterEventListener.java [hunk 7]
```diff
@@ -24,16 +29,32 @@ public class MLCommonsClusterEventListener implements ClusterStateListener {
...
     @Override
     public void clusterChanged(ClusterChangedEvent event) {
+        ClusterState previousState = event.previousState();
+        ClusterState state = event.state();
+        Settings previousSettings = previousState.getMetadata().settings();
+        Settings settings = state.getMetadata().settings();
+        long previousMonitoringReqCount = ML_COMMONS_MONITORING_REQUEST_COUNT.get(previousSettings);
+        long monitoringReqCount = ML_COMMONS_MONITORING_REQUEST_COUNT.get(settings);
+        if (previousMonitoringReqCount > monitoringReqCount) {
+            modelCacheHelper.resizeMonitoringQueue(monitoringReqCount);
```
Comment (jngz-es) [inline plugin/src/main/java/org/opensearch/ml/cluster/MLCommonsClusterEventListener.java:56]: Do we have a maximum size limitation on this queue?

Comment (ylwu-amzn) [reply plugin/src/main/java/org/opensearch/ml/cluster/MLCommonsClusterEventListener.java:56]: yes, the line above contains the limie "monitoringReqCount"

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java (hunk 6 above)
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/text_embedding/TextEmbeddingModel.java:270]: Good point, sounds good, will change this part.

**Pull request merged**